### PR TITLE
Pick smallest decimal type with required precision in ORC reader

### PR DESCRIPTION
**Pull request by vuule**
Depends on #9853 

Current behavior is to throw when an ORC column has precision that is too high for decimal64. 
This PR changes the behavior to instead read the column as decimal128, when precision is too high for 64 bits. This reduces the need for the use of `decimal128_columns` option.
Also modified the decimal type inference to use decimal32 when the precision is sufficiently low, reducing memory use in such case.
Adds a temporary option to disable decimal128 use. This option is used in Python to get a readable error message in this case, while allowing decimal128 use by other callers.

### CODE REVIEW

**Comment by vuule** [comment]
@galipremsagar please review behavior when calling from Python. Since decimal128 is disabled through the new option, users shoudl get a nice error message when reading columns with too high precision, and have an option to read as float (no changes there).

**Review by galipremsagar — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

File: cpp/src/io/orc/reader_impl.hpp [hunk 1]
```diff
@@ -223,6 +223,7 @@ class reader::impl {
   bool _use_np_dtypes = true;
   std::vector<std::string> _decimal_cols_as_float;
   std::vector<std::string> decimal128_columns;
+  bool is_decimal128_enabled;
   data_type _timestamp_type{type_id::EMPTY};
   reader_column_meta _col_meta;
```
Comment (harrism) [inline cpp/src/io/orc/reader_impl.hpp:228]: Inconsistency in what is initialized and how things are initialized here. Two `bool` members above are initialized to true using `=`. This `bool` is not initialized. `data_type` is initialized using `{}`. Suggest adding initialization for everything and doing them all the same way.

**Review by harrism — APPROVED** [review]
Looks good. Just one comment.

**Comment by hyperbolic2346** [comment]
Is there a performance penalty if reading decimal128 data?

File: cpp/src/io/orc/reader_impl.hpp (hunk 1 above)
Comment (vuule) [reply cpp/src/io/orc/reader_impl.hpp]: Done, only left the vectors without "visible" initialization.

**Comment by vuule** [comment]
> Is there a performance penalty if reading decimal128 data?

Do you mean reading the same column as decimal128 as opposed to decimal32/64? Shouldn't be a big difference (just larger stride between writes), but I don't have a benchmark set up for this.

What I could do with existing benchmarks is to read N megabytes of decimal32 vs 64 vs 128. Here, decimal128 Is much faster than decimal32/64. Performance is almost proportional to the type width in this case. It's possible that the random number generator does not generate fair data, but at least it does not look like there's an inherent penalty from using decimal128.

**Review by hyperbolic2346 — APPROVED** [review]
Reading the initial description I thought this was a multi-pass algorithm if the data was decimal128 or decimal32. I was worried about the performance implications, but I see there is metadata for the precision. Looks good to me.

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**